internal/command/run.go

```diff
@@ -184,77 +197,120 @@ func (c *runCmd) run(cmd *cobra.Command, args []string) {
 		ptCopy := pt
 		c.taskRunnerRoutinePool.Queue(func() {
 			task := ptCopy.task
-			runResult := c.runTask(task)
+			runResult, err := c.runTask(task)
+			if err != nil {
+				// error is printed in runTask()
+				c.terminateGracefully()
+				return
+			}
 
 			outputs, err := baur.OutputsFromTask(c.dockerClient, task)
-			exitOnErrf(err, task.ID())
+			if err != nil {
+				stderr.ErrPrintln(err, task.ID())
+				c.terminateGracefully()
+				return
+			}
 
 			if !outputsExist(task, outputs) {
-				exitFunc(1)
+				// error is printed in runTask()
+				c.terminateGracefully()
+				return
 			}
 
 			if c.skipUpload {
 				return
 			}
 
 			c.uploadRoutinePool.Queue(func() {
-				c.uploadAndRecord(ctx, ptCopy.task, ptCopy.inputs, outputs, runResult)
+				err := c.uploadAndRecord(ctx, ptCopy.task, ptCopy.inputs, outputs, runResult)
+				if err != nil {
+					// error is printed in runTask()
+					c.terminateGracefully()
+				}
 			})
 		})
 	}
 
 	c.taskRunnerRoutinePool.Wait()
 
-	stdout.Println("all tasks executed, waiting for uploads to finish...")
+	stdout.Println("task execution finished, waiting for uploads to finish...")
 	c.uploadRoutinePool.Wait()
 	stdout.PrintSep()
 	stdout.Printf("finished in: %s\n",
 		term.FormatDuration(
 			time.Since(startTime),
 		),
 	)
+
+	if c.errorHappened {
+		exitFunc(1)
+	}
 }
 
-func (c *runCmd) runTask(task *baur.Task) *baur.RunResult {
-	result, err := baur.NewTaskRunner().Run(task)
-	exitOnErrf(err, "%s", task.ID())
+func (c *runCmd) terminateGracefully() {
```

Comment:
> Don't this method have to be synchronous (aka wait for tasks to finish)?

That is done in `run()`.
The function is called from the go-routines itself, so it can't wait until they terminated.

Yes, the name was unclear.
I changed it to `c.skipAllScheduledTaskRuns()`.
Thanks!